### PR TITLE
add multi-item scoring

### PR DESCRIPTION
**Pull request by arde171**
Co-authored with Qingquan Song (@qingquansong) and Ziang Li (@zianglih )

**Multi-item scoring** 
1. concatenate multiple candidates of a same member with all ranking candidates with delimiter separation.
<member prefix (profile & history)> + <delimiter> + <item1> + <delimiter> + item 2 + ... + item N <delimiter>
2. Extract the logits of the hidden states of the tokens before each delimiter token and extract the log prob of given label tokens. For each single prompt, output returned will be a 2D list with shape N * K where N is the number of candidate it contains and K is the number of choices we provided to the server engine (e.g., 2 for ["Yes", "No"])) (mainly done in the logit processor)
image

The PR optimized the multi-item scoring attention by passing four new args and use it to check the masking condition. The provided args are:
```
prefix_len_ptr :Optional[torch.Tensor]
    prefix length. A uint32 1D tensor indicating the prefix length of each prompt. The tensor size is equal to the batch size.
token_pos_in_items_ptr : Optional[float]
    A uint16 1D tensor (it will be converted to uint16 in flashinfer) indicating the token position of each item and started from 0 (delimiter)
    for each item. E.g., if we have 3 items of length 3, 2, 4 respectively for this member. This vector will be looking like
    `[0, 1, 2, 3, 0, 1, 2, 0, 1, 2, 3, 4, 0]` with 4 delimiters indexed as 0. For batch size > 1,
    we will concat them as 1D with zero paddings to make sure each has the same length, the padding length is defined by
    `token_pos_in_items_len` - length of the raw `token_pos_in_items_ptr` for each prompt.
token_pos_in_items_len : Optional[int]
    zero padding length for `token_pos_in_items_ptr` to better handle the bsz > 1 case. Still using the above 3,2,4 example.
    If we set `token_pos_in_items_len` to be 20, it will be  `[0, 1, 2, 3, 0, 1, 2, 0, 1, 2, 3, 4, 0, 0, 0, 0, 0, 0, 0, 0]`
    with 7 padded zeros. (note there're 8 zeros in the end where the first one is the delimiter token 0 in the end of the prompt)
max_item_len_ptr : Optional[float]
    a uint16 vector contains the max token length of all items for each prompt
```

**Optimizations**
1. Implement efficient multi-item scoring mask for FA2 and FA3.
2. Enhance FA3  to support batch-idx for the multi-item scoring mask.
3. Implement skip tiles for FA2 and FA3 multi-item scoring
4. Optimize mask by preloading to L1 cache for thread register.

### CODE REVIEW

**Comment by qingquansong** [comment]
Hey @yzh119 as discussed, here's the PR for multi-item scoring masked attention. Please feel free to leave comments and provide suggestions if there could be better ways to help upstream the change. Thank you in advance!

File: csrc/batch_prefill_customize_config.jinja [hunk 1]
```diff
@@ -108,6 +112,10 @@ struct PagedParams {
   uint32_t* total_num_rows;
   uint32_t padded_batch_size;
   bool partition_kv;
+  uint32_t* prefix_len_ptr;
```
Comment (yzh119) [inline csrc/batch_prefill_customize_config.jinja]: Can you move them to additional params? I tend to managing all of the options parameters as additional, instead of default ones, which is easier to manager.

Examples include: https://github.com/flashinfer-ai/flashinfer/blob/9220fb3443b5a5d274f00ca5552f798e225239b7/flashinfer/jit/attention/pytorch.py#L456 (we use prefix `maybe_` to indicate these components are optional and have type std::optional<...>).

File: csrc/batch_prefill_sm90_customize_config.jinja [hunk 2]
```diff
@@ -66,6 +67,11 @@ struct RaggedParams {
   int window_left;
 
   bool causal;
+
```
Comment (yzh119) [inline csrc/batch_prefill_sm90_customize_config.jinja]: Ditto, better to move to additional params.

File: csrc/batch_prefill_sm90_customize_config.jinja [hunk 3]
```diff
@@ -43,6 +43,7 @@ struct RaggedParams {
   IdType* kv_lens;
   IdType* head_indices;
   IdType* work_indptr;
+  IdType* batch_indices;
```
Comment (yzh119) [inline csrc/batch_prefill_sm90_customize_config.jinja:46]: Thanks for doing this, yes we have to add it.

**Review by yzh119 — COMMENTED** [review]
Overall LGTM, leave some comments on additional parameters.

btw, some unittests failed (https://github.com/flashinfer-ai/flashinfer/blob/9220fb3443b5a5d274f00ca5552f798e225239b7/tests/test_block_sparse.py) bacause of the pybind interface change, would you mind fixing the sparse APIs as well (https://github.com/flashinfer-ai/flashinfer/blob/main/flashinfer/sparse.py)?

File: include/flashinfer/attention/prefill.cuh [hunk 4]
```diff
@@ -786,6 +786,73 @@ __device__ __forceinline__ void logits_mask(
   }
 }
 
+template <typename KTraits, typename Params>
+__device__ __forceinline__ void logits_mask_customized(
```
Comment (zianglih) [inline include/flashinfer/attention/prefill.cuh]: Better rename this function to "logits_mask_multi_item_scoring". Or move its body to the previous "logits_mask".

File: include/flashinfer/attention/prefill.cuh [hunk 5]
```diff
@@ -2093,13 +2165,43 @@ __device__ __forceinline__ void BatchPrefillWithPagedKVCacheDevice(
...
+      num_iterations =
+          ceil_div((MASK_MODE == MaskMode::kCausal
+                        ? min(chunk_size,
+                              sub_if_greater_or_zero(
+                                  kv_len - qo_len + ((qo_tile_idx + 1) * CTA_TILE_Q) / group_size,
+                                  chunk_start))
+                        : chunk_size),
+                   CTA_TILE_KV);
+    } else {
+      num_iterations = ceil_div(
```
Comment (zianglih) [inline include/flashinfer/attention/prefill.cuh:2182]: Maybe should rename MIS "num_iterations" to "num_iterations_prefix".

File: include/flashinfer/attention/prefill.cuh [hunk 6]
```diff
@@ -2114,9 +2216,18 @@ __device__ __forceinline__ void BatchPrefillWithPagedKVCacheDevice(
              : chunk_size) /
         CTA_TILE_KV;
 
+    const uint32_t unified_num_iterations =
```
Comment (zianglih) [inline include/flashinfer/attention/prefill.cuh]: Maybe rename MIS "num_iterations_full" to "num_iterations", and MIS "num_iterations" to "num_iterations_prefix" to avoid redundancy.

File: csrc/batch_prefill_customize_config.jinja (hunk 1 above)
Comment (arde171) [reply csrc/batch_prefill_customize_config.jinja]: @yzh119 as suggested, moved multi-item scoring parameters as addtional.

File: csrc/batch_prefill_sm90_customize_config.jinja (hunk 2 above)
Comment (arde171) [reply csrc/batch_prefill_sm90_customize_config.jinja]: Done.

File: include/flashinfer/attention/prefill.cuh (hunk 4 above)
Comment (arde171) [reply include/flashinfer/attention/prefill.cuh]: fixed

File: include/flashinfer/attention/prefill.cuh (hunk 6 above)
Comment (arde171) [reply include/flashinfer/attention/prefill.cuh]: Done.

**Comment by qingquansong** [comment]
Hey @yzh119 , @arde171 has resolved the comments, could you help take another look? Thank you!

**Review by yzh119 — APPROVED** [review]
Hi @arde171 @qingquansong @zianglih thanks for the great contribution, the PR looks good to me in general and I have add some commits to address the conflicts with mainline. Let's merge this PR first and then move forward.

The remaining possible improvements include (not necessarily in this PR):
1. consider the skipped blocks in scheduler (`plan` function), otherwise the ahead-of-time scheduler might have false estimation about execution time of a tile.
2. further modularize the template to make the attention pattern in multi-item scoring a special form of attention variant, currently we still insert some special code to handle this pattern in the template but we hope we can fully decouple attention variant from the template themselves.
